### ISSUE TITLE
Null Object Reference - DeepClone

### ISSUE THREAD

**Issue by vsederburg** [agileobjects/AgileMapper#209]
I'm getting an error using Mapper.DeepClone to clone a class. This was working sometime in the past few weeks, so I'm not sure what has changed. I'm using the current Mapper version 1.7.2 via nuGet.

I seem to remember at one point there was a problem cloning an object that included a dictionary, but I think that was fixed. 

I've attached the class and base class that I'm trying to clone. Let me know if I can provide any other information.

Thank you!

This is the error:

System.NullReferenceException: Object reference not set to an instance of an object.
   at AgileObjects.AgileMapper.Members.QualifiedMember.Append(Member childMember)
   at AgileObjects.AgileMapper.Members.MemberMapperDataExtensions.GetNonEnumerableChildMember(QualifiedMember parentMember, Member childMember)
   at AgileObjects.AgileMapper.Members.MemberMapperDataExtensions.TargetMemberEverRepeatsWithin(QualifiedMember parentMember, IQualifiedMember subjectMember)
   at AgileObjects.AgileMapper.Members.MemberMapperDataExtensions.TargetMemberEverRepeatsWithin(QualifiedMember parentMember, IQualifiedMember subjectMember)
   at AgileObjects.AgileMapper.Members.MemberMapperDataExtensions.TargetMemberEverRepeatsWithin(QualifiedMember parentMember, IQualifiedMember subjectMember)
   at AgileObjects.AgileMapper.Members.MemberMapperDataExtensions.TargetMemberEverRepeatsWithin(QualifiedMember parentMember, IQualifiedMember subjectMember)
   at AgileObjects.AgileMapper.Members.MemberMapperDataExtensions.TargetMemberEverRepeatsWithin(QualifiedMember parentMember, IQualifiedMember subjectMember)
   at AgileObjects.AgileMapper.Members.MemberMapperDataExtensions.TargetMemberEverRepeatsWithin(QualifiedMember parentMember, IQualifiedMember subjectMember)
   at AgileObjects.AgileMapper.Members.MemberMapperDataExtensions.TargetMemberEverRepeatsWithin(QualifiedMember parentMember, IQualifiedMember subjectMember)
   at AgileObjects.AgileMapper.Members.MemberMapperDataExtensions.TargetMemberEverRepeatsWithin(QualifiedMember parentMember, IQualifiedMember subjectMember)
   at AgileObjects.AgileMapper.Members.MemberMapperDataExtensions.TargetMemberEverRepeatsWithin(QualifiedMember parentMember, IQualifiedMember subjectMember)
   at AgileObjects.AgileMapper.Members.MemberMapperDataExtensions.IsRepeatMapping(IQualifiedMemberContext context)
   at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapperData.get_IsRepeatMapping()
   at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapperData.IsNotTargetTypeFirstMapping(ObjectMapperData parent)
   at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapperData..ctor(IMappingContext mappingContext, IQualifiedMember sourceMember, QualifiedMember targetMember, Nullable`1 dataSourceIndex, ObjectMapperData declaredTypeMapperData, ObjectMapperData parent, Boolean isForStandaloneMapping)
   at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapperData.For[TSource,TTarget](ObjectMappingData`2 mappingData)
   at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingData`2.get_MapperData()
   at AgileObjects.AgileMapper.ObjectPopulation.MappingFactory.GetChildMapping(IObjectMappingData childMappingData, MappingValues mappingValues, ObjectMapperData declaredTypeMapperData)
   at AgileObjects.AgileMapper.ObjectPopulation.MappingFactory.GetChildMapping(IQualifiedMember sourceMember, Expression sourceMemberAccess, Int32 dataSourceIndex, IChildMemberMappingData childMappingData)
   at AgileObjects.AgileMapper.DataSources.EnumerableDataSource.GetMapping(IDataSource sourceEnumerableDataSource, Int32 dataSourceIndex, IChildMemberMappingData enumerableMappingData)
   at AgileObjects.AgileMapper.DataSources.Factories.DataSourceFindContext.GetFinalDataSource(IDataSource foundDataSource, IChildMemberMappingData mappingData)
   at AgileObjects.AgileMapper.DataSources.Factories.DataSourceFindContext.GetSourceMemberDataSource()
   at AgileObjects.AgileMapper.DataSources.Factories.DataSourceFindContext.get_MatchingSourceMemberDataSource()
   at AgileObjects.AgileMapper.DataSources.Factories.DataSourceFindContext.UseSourceMemberDataSource()
   at AgileObjects.AgileMapper.DataSources.Factories.SourceMemberDataSourcesFactory.<Create>d__0.MoveNext()
   at AgileObjects.AgileMapper.DataSources.Factories.MemberDataSourceSetFactory.<EnumerateDataSources>d__2.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at AgileObjects.AgileMapper.DataSources.Factories.MemberDataSourceSetFactory.CreateFor(DataSourceFindContext findContext)
   at AgileObjects.AgileMapper.Members.Population.MemberPopulatorFactory.Create(MemberPopulationContext context)
   at AgileObjects.AgileMapper.Members.Population.MemberPopulatorFactory.<>c.<Create>b__3_0(MemberPopulationContext ctx, QualifiedMember tm)
   at AgileObjects.AgileMapper.Extensions.PublicEnumerableExtensions.<Project>d__1`3.MoveNext()
   at AgileObjects.AgileMapper.Extensions.PublicEnumerableExtensions.<Filter>d__4`2.MoveNext()
   at AgileObjects.AgileMapper.ObjectPopulation.ComplexTypes.PopulationExpressionFactoryBase.<GetPopulationsAndCallbacks>d__3.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at AgileObjects.AgileMapper.ObjectPopulation.ComplexTypes.PopulationExpressionFactoryBase.AddPopulation(MappingCreationContext context)
   at AgileObjects.AgileMapper.ObjectPopulation.ComplexTypes.ComplexTypeMappingExpressionFactory.AddObjectPopulation(MappingCreationContext context)
   at AgileObjects.AgileMapper.ObjectPopulation.MappingExpressionFactoryBase.<AddPopulationsAndCallbacks>b__11_0(MappingExpressionFactoryBase factory, MappingCreationContext ctx)
   at AgileObjects.AgileMapper.ObjectPopulation.MappingExpressionFactoryBase.AddPopulationsAndCallbacks[TArg](TArg argument, MappingCreationContext context, Action`2 mappingBodyPopulator)
   at AgileObjects.AgileMapper.ObjectPopulation.MappingExpressionFactoryBase.AddPopulationsAndCallbacks(MappingCreationContext context)
   at AgileObjects.AgileMapper.ObjectPopulation.MappingExpressionFactoryBase.Create(IObjectMappingData mappingData)
   at AgileObjects.AgileMapper.DataSources.ComplexTypeDataSource.Create(IObjectMappingData mappingData)
   at AgileObjects.AgileMapper.DataSources.Factories.Mapping.ComplexTypeMappingDataSourceFactory.CreateFor(IObjectMappingData mappingData)
   at AgileObjects.AgileMapper.DataSources.Factories.Mapping.MappingDataSourceSetFactory.CreateFor(IObjectMappingData mappingData)
   at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapperFactory.Create[TSource,TTarget](ObjectMappingData`2 mappingData)
   at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingData`2.GetOrCreateMapper()
   at AgileObjects.AgileMapper.ObjectPopulation.MappingFactory.GetInlineMappingBlock(IObjectMappingData mappingData, MappingValues mappingValues, CreateMappingDataCallFactory createMappingDataCallFactory)
   at AgileObjects.AgileMapper.ObjectPopulation.MappingFactory.GetElementMapping(IObjectMappingData elementMappingData, Expression sourceElementValue, Expression targetElementValue)
   at AgileObjects.AgileMapper.ObjectPopulation.MappingFactory.GetElementMapping(Expression sourceElementValue, Expression targetElementValue, IObjectMappingData mappingData)
   at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.EnumerablePopulationBuilder.GetElementConversion(Expression sourceElement, IObjectMappingData enumerableMappingData)
   at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.Looping.IndexedSourcePopulationLoopData.GetElementMapping(IObjectMappingData enumerableMappingData)
   at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.EnumerablePopulationBuilder.GetElementPopulation(IPopulationLoopData loopData, IObjectMappingData enumerableMappingData)
   at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.Looping.PopulationLoopDataExtensions.BuildPopulationLoop[TLoopData](TLoopData loopData, EnumerablePopulationBuilder builder, IObjectMappingData mappingData, Func`3 elementPopulationFactory)
   at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.EnumerablePopulationBuilder.BuildPopulationLoop(Func`3 elementPopulationFactory, IObjectMappingData mappingData)
   at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.EnumerablePopulationBuilder.AddNewItemsToTargetVariable(IObjectMappingData enumerableMappingData)
   at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.CopySourceEnumerablePopulationStrategy.Create(EnumerablePopulationBuilder builder, IObjectMappingData enumerableMappingData)
   at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.EnumerableMappingExpressionFactory.AddObjectPopulation(MappingCreationContext context)
   at AgileObjects.AgileMapper.ObjectPopulation.MappingExpressionFactoryBase.<AddPopulationsAndCallbacks>b__11_0(MappingExpressionFactoryBase factory, MappingCreationContext ctx)
   at AgileObjects.AgileMapper.ObjectPopulation.MappingExpressionFactoryBase.AddPopulationsAndCallbacks[TArg](TArg argument, MappingCreationContext context, Action`2 mappingBodyPopulator)
   at AgileObjects.AgileMapper.ObjectPopulation.MappingExpressionFactoryBase.AddPopulationsAndCallbacks(MappingCreationContext context)
   at AgileObjects.AgileMapper.ObjectPopulation.MappingExpressionFactoryBase.Create(IObjectMappingData mappingData)
   at AgileObjects.AgileMapper.DataSources.Factories.Mapping.MappingDataSourceFactoryBase.CreateFor(IObjectMappingData mappingData)
   at AgileObjects.AgileMapper.DataSources.Factories.Mapping.MappingDataSourceSetFactory.CreateFor(IObjectMappingData mappingData)
   at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapperFactory.Create[TSource,TTarget](ObjectMappingData`2 mappingData)
   at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingData`2.GetOrCreateMapper()
   at AgileObjects.AgileMapper.ObjectPopulation.MappingFactory.GetInlineMappingBlock(IObjectMappingData mappingData, MappingValues mappingValues, CreateMappingDataCallFactory createMappingDataCallFactory)
   at AgileObjects.AgileMapper.ObjectPopulation.MappingFactory.GetChildMapping(IObjectMappingData childMappingData, MappingValues mappingValues, ObjectMapperData declaredTypeMapperData)
   at AgileObjects.AgileMapper.ObjectPopulation.MappingFactory.GetChildMapping(IQualifiedMember sourceMember, Expression sourceMemberAccess, Int32 dataSourceIndex, IChildMemberMappingData childMappingData)
   at AgileObjects.AgileMapper.DataSources.EnumerableDataSource.GetMapping(IDataSource sourceEnumerableDataSource, Int32 dataSourceIndex, IChildMemberMappingData enumerableMappingData)
   at AgileObjects.AgileMapper.DataSources.Factories.DataSourceFindContext.GetFinalDataSource(IDataSource foundDataSource, IChildMemberMappingData mappingData)
   at AgileObjects.AgileMapper.DataSources.Factories.DataSourceFindContext.GetSourceMemberDataSource()
   at AgileObjects.AgileMapper.DataSources.Factories.DataSourceFindContext.get_MatchingSourceMemberDataSource()
   at AgileObjects.AgileMapper.DataSources.Factories.DataSourceFindContext.UseSourceMemberDataSource()
   at AgileObjects.AgileMapper.DataSources.Factories.SourceMemberDataSourcesFactory.<Create>d__0.MoveNext()
   at AgileObjects.AgileMapper.DataSources.Factories.MemberDataSourceSetFactory.<EnumerateDataSources>d__2.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at AgileObjects.AgileMapper.DataSources.Factories.MemberDataSourceSetFactory.CreateFor(DataSourceFindContext findContext)
   at AgileObjects.AgileMapper.ObjectPopulation.ComplexTypes.ComplexTypeConstructionFactory.ConstructionDataInfo`1..ctor(TInvokable invokable, ConstructionKey key, Int32 priority)
   at AgileObjects.AgileMapper.ObjectPopulation.ComplexTypes.ComplexTypeConstructionFactory.ObjectNewingInfo..ctor(ConstructorInfo ctor, ConstructionKey key)
   at AgileObjects.AgileMapper.ObjectPopulation.ComplexTypes.ComplexTypeConstructionFactory.<>c__DisplayClass9_0.<GetGreediestAvailableNewingInfos>b__0(ConstructorInfo ctor)
   at AgileObjects.AgileMapper.Extensions.PublicEnumerableExtensions.<>c__0`2.<Project>b__0_0(Func`2 p, TItem item)
   at AgileObjects.AgileMapper.Extensions.PublicEnumerableExtensions.<Project>d__1`3.MoveNext()
   at AgileObjects.AgileMapper.Extensions.PublicEnumerableExtensions.<Filter>d__4`2.MoveNext()
   at AgileObjects.AgileMapper.Extensions.Internal.EnumerableExtensions.<TakeUntil>d__6`1.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at AgileObjects.AgileMapper.ObjectPopulation.ComplexTypes.ComplexTypeConstructionFactory.CreateConstructionInfo[T](IEnumerable`1 invokables, Func`2 dataFactory)
   at AgileObjects.AgileMapper.ObjectPopulation.ComplexTypes.ComplexTypeConstructionFactory.GetGreediestAvailableNewingInfos(IEnumerable`1 constructors, ConstructionKey key, IBasicConstructionInfo greediestUnconditionalFactoryInfo)
   at AgileObjects.AgileMapper.ObjectPopulation.ComplexTypes.ComplexTypeConstructionFactory.AddAutoConstructionInfos(IList`1 constructionInfos, ConstructionKey key)
   at AgileObjects.AgileMapper.ObjectPopulation.ComplexTypes.ComplexTypeConstructionFactory.<>c.<GetTargetObjectCreationInfos>b__4_0(ConstructionKey key)
   at AgileObjects.AgileMapper.Caching.ArrayCache`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at AgileObjects.AgileMapper.ObjectPopulation.ComplexTypes.ComplexTypeConstructionFactory.GetTargetObjectCreationInfos(IObjectMappingData mappingData, ConstructionKey& constructionKey)
   at AgileObjects.AgileMapper.ObjectPopulation.ComplexTypes.ComplexTypeConstructionFactory.GetTargetObjectCreationInfos(IObjectMappingData mappingData)
   at AgileObjects.AgileMapper.MappingDataExtensions.GetTargetObjectCreationInfos(IObjectMappingData mappingData)
   at AgileObjects.AgileMapper.ObjectPopulation.ComplexTypes.ComplexTypeMappingExpressionFactory.TargetCannotBeMapped(IObjectMappingData mappingData, String& reason)
   at AgileObjects.AgileMapper.ObjectPopulation.MappingExpressionFactoryBase.Create(IObjectMappingData mappingData)
   at AgileObjects.AgileMapper.DataSources.ComplexTypeDataSource.Create(IObjectMappingData mappingData)
   at AgileObjects.AgileMapper.DataSources.Factories.Mapping.ComplexTypeMappingDataSourceFactory.CreateFor(IObjectMappingData mappingData)
   at AgileObjects.AgileMapper.DataSources.Factories.Mapping.MappingDataSourceSetFactory.CreateFor(IObjectMappingData mappingData)
   at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapperFactory.Create[TSource,TTarget](ObjectMappingData`2 mappingData)
   at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingData`2.GetOrCreateMapper()
   at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapperFactory.<>c__6`2.<GetOrCreateRoot>b__6_0(IRootMapperKey key)
   at AgileObjects.AgileMapper.Caching.ArrayCache`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapperFactory.GetOrCreateRoot[TSource,TTarget](ObjectMappingData`2 mappingData)
   at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingData`2..ctor(TSource source, TTarget target, Nullable`1 elementIndex, Object elementKey, MappingTypes mappingTypes, IMappingContext mappingContext, IObjectMappingData declaredTypeMappingData, IObjectMappingData parent, Boolean createMapper)
   at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingDataFactory.ForRootFixedTypes[TSource,TTarget](TSource source, TTarget target, MappingTypes mappingTypes, IMappingContext mappingContext, Boolean createMapper)
   at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingDataFactory.ForRootFixedTypes[TSource,TTarget](TSource source, TTarget target, IMappingContext mappingContext, Boolean createMapper)
   at AgileObjects.AgileMapper.MappingExecutor`1.PerformMapping[TTarget](TTarget target)
   at AgileObjects.AgileMapper.MappingExecutor`1.PerformMapping[TTarget](MappingRuleSet ruleSet, TTarget target)
   at AgileObjects.AgileMapper.MappingExecutor`1.ToANew[TResult]()
   at AgileObjects.AgileMapper.Mapper.AgileObjects.AgileMapper.IMapper.DeepClone[TSource](TSource source)
   at AgileObjects.AgileMapper.Mapper.DeepClone[TSource](TSource source)
   at mt.Client.Common.ViewModels.Activity.AttendanceGridSettingsViewModel.OnInitialize(Object[] args) in C:\Data\mt_src\csharp\mt\mt.Client.Common\ViewModels\Activity\AttendanceGridSettingsViewModel.cs:line 98
   at mt.Shared.Base.MtViewModelBase.Initialize(IMtViewModelParent parent, Object[] args) in C:\Data\mt_src\csharp\mt\mt.Shared\Base\MtViewModelBase.cs:line 93
   at mt.Shared.Base.MtViewModelBase.Initialize(IMtViewModelParent parent) in C:\Data\mt_src\csharp\mt\mt.Shared\Base\MtViewModelBase.cs:line 76
   at mt.Client.Common.UserControls.Activity.AttendanceGridSettingsUserControl.InitializeStart(Object[] args) in C:\Data\mt_src\csharp\mt\mt.Client.Common\UserControls\Activity\AttendanceGridSettingsUserControl.cs:line 49
   at mt.Shared.Base.MtUserControlBase.Initialize(Object[] args) in C:\Data\mt_src\csharp\mt\mt.Shared\Base\MtUserControlBase.cs:line 148
2021-03-09 08:02:36,278 [1] INFO - Notification: Object reference not set to an instance of an object.

[ViewDataDto.cs.txt](https://github.com/agileobjects/AgileMapper/files/6109078/ViewDataDto.cs.txt)

[ViewDataGridPreferencesDto.cs.txt](https://github.com/agileobjects/AgileMapper/files/6109072/ViewDataGridPreferencesDto.cs.txt)

**Comment by vsederburg** [comment]
Oh, and I'm calling the clone simply with:

private readonly ViewDataGridPreferencesDto _viewData; // set in constructor
private ViewDataGridPreferencesDto _originalViewData;

protected some method
{
_originalViewData = Mapper.DeepClone(_viewData);
}

**Comment by SteveWilkes** [comment]
Hello!

Thanks for letting me know about this - could I have the source for `Preference`, `UserPreferenceDto` and any other concrete types in the object tree?

Cheers,

Steve

**Comment by SteveWilkes** [comment]
Unfortunately not! You could attach them here or make a Git repo with them in?

**Comment by vsederburg** [comment]
Files attached!

[files.zip](https://github.com/agileobjects/AgileMapper/files/6122934/files.zip)

**Comment by SteveWilkes** [comment]
Thanks very much - I'll check this out as soon as I can (y)

**Comment by replaysMike** [comment]
I'm not sure if it's related, but we are seeing null ref's internally occurring in `Map<TDest>(TSource)` when mapping ICollections and not sure if it's related to the above or not yet as I can't yet repro it in a unit test, but it happens in our pipeline.

**Comment by replaysMike** [comment]
my comment above was moved to it's own issue as it's different: #212

**Comment by SteveWilkes** [comment]
Hi @vsederburg,

Just looking at this now - could you let me know where the `ObservableObject` type from which your `DtoBase` tyoe derives is from? I can't find one in the framework in any of the namespaces listed  in `DtoBase`'s source...

Also - and this isn't the cause of this issue and you may already have found it, but just in case - to deep-clone a `ViewDataGridPreferencesDto` you'll have to tell the mapper what to use for the `callerSource` argument as the corresponding property is named `Source`:

```csharp
_originalViewData = Mapper.DeepClone(_viewData, cfg => cfg
    .Map((svd, tvd) => svd.Source)
    .ToCtor<GridPreferenceCallerSource>());
```

**Comment by vsederburg** [comment]
I've attached the ObservableObject class. Sorry I missed that one.

For the ViewDataGridPreferencesDto you mentioned... would it fix that if I just changed the corresponding property name to CallerSource, or make them match one way or the other?

[ObservableObject.cs.txt](https://github.com/agileobjects/AgileMapper/files/6264669/ObservableObject.cs.txt)

**Comment by SteveWilkes** [comment]
Hi,

Thanks for `ObservableObject`'s source - the mapper believe it or not was falling over trying to figure out how to map to its `PropertyInfo[]`, so I've updated target member selection to exclude most BCL non-value-type targets. The mapper isn't really intended to be used to map `PropertyInfo`s or `Stream`s or `HttpContext`s or anything like that, so I believe it will still cover the vast majority of scenarios.

Thanks again for reporting the issue!

All the best,

Steve

**Comment by vsederburg** [comment]
That's weird, that has been the base class forever and it's never complained before.   Oh well, looking forward to the update. This mapper is super useful. :-)

**Comment by SteveWilkes** [comment]
Hi,

I've now uploaded a preview v1.8 release [to NuGet](https://www.nuget.org/packages/AgileObjects.AgileMapper) with this issue fixed - please could you try it out and let me know how you get on?

Thanks,

Steve

**Comment by vsederburg** [comment]
YES! All seems to be working again. Thanks so much!

**Comment by SteveWilkes** [comment]
Excellent! You're welcome, thanks for checking :)

**Comment by SteveWilkes** [comment]
This fix is in the full v1.8 release, which is now available [on NuGet](https://www.nuget.org/packages/AgileObjects.AgileMapper).

Thanks!
